### ISSUE TITLE
Clarify the behavior of the default output validator and the semantics of the default output validator flags

### ISSUE THREAD

**Issue by evouga** [Kattis/problem-package-format#348]
The behavior of the default output validator is not well specified.

1. What output encoding does it require? UTF-8 without BOM? How is the output tokenized? How is the "type" of the token (floating-point vs. string, I guess? Maybe integer too, see below?) determined?

2. Are integers compared as strings? (With leading zeros, hexadecimal encoding, etc. being judged as incorrect) or is there some special handling for "integer-type" tokens?

3. `case_sensitive`: notoriously [underspecified](https://www.b-list.org/weblog/2018/nov/26/case/). How exactly is case-insensitive comparison performed? I assume that this flag only applies to tokens of "string type" and not floating-point?

4. `space_change_sensitive`: The default behavior is mentioned in a parenthetical remark `(the default is that any sequence of 1 or more whitespace characters are equivalent)` but this is incomplete since it does not mention leading or trailing whitespace (which I understand the current default validator ignores by default). Moreover `indicates that changes in the amount of whitespace should be rejected` is imprecise: do changes in the *type* of whitespace (newlines to spaces e.g.) cause submissions to be rejected? What is a whitespace anyway? The 25 whitespace Unicode code points?

**Comment by Tagl** [comment]
I agree this needs a better explanation.

> 1. What output encoding does it require? UTF-8 without BOM? How is the output tokenized? How is the "type" of the token (floating-point vs. string, I guess? Maybe integer too, see below?) determined?

The current behavior is: if it can be parsed as a float, and is not nan or inf, then it is a float. This is obviously not great because integers can be parsed as floats and although you might want to allow precision issues with `3.1416` you might not want it for `3`.

> 2. Are integers compared as strings? (With leading zeros, hexadecimal encoding, etc. being judged as incorrect) or is there some special handling for "integer-type" tokens?

If float tolerance is set, then the answer for 1 applies, but otherwise string comparison is used. I'd argue that we want to change the behavior of 1 so integers use string comparison always. Do you oppose string comparison for integers?

> 3\. `case_sensitive`: notoriously [underspecified](https://www.b-list.org/weblog/2018/nov/26/case/). How exactly is case-insensitive comparison performed? I assume that this flag only applies to tokens of "string type" and not floating-point?

When this was introduced, I believe only ASCII was supported, at least that was my experience. Currently, it is just ASCII case-insensitivity that is applied using `strcasecmp`. Yes, it is true that casing is more complicated than upper and lower case. If we do make the default validator fully support UTF-8 then it should probably have flags for case-sensitive, accent-sensitive, kana-sensitive, width-sensitive, variation selector sensitive. If floating tolerance is provided then floats are not string compared.

> 4\. `space_change_sensitive`: The default behavior is mentioned in a parenthetical remark `(the default is that any sequence of 1 or more whitespace characters are equivalent)` but this is incomplete since it does not mention leading or trailing whitespace (which I understand the current default validator ignores by default). Moreover `indicates that changes in the amount of whitespace should be rejected` is imprecise: do changes in the _type_ of whitespace (newlines to spaces e.g.) cause submissions to be rejected? What is a whitespace anyway? The 25 whitespace Unicode code points?

The whitespace tokens are currently the ones for which [std::isspace](https://en.cppreference.com/w/cpp/string/byte/isspace) returns true. Should probably add the remaining unicode code points. Changes in types of white space are also ignored, the only thing that matters is that there is some separation of the tokens so that the can be distinguished as separate tokens and not the same token. Therefore trailing and leading do not matter either.

**Comment by RagnarGrootKoerkamp** [comment]
I would be careful to disable float comparison for integer-looking values: if the the floating-point output to some problems happens to be exactly `1000000`, then suddenly a team answer of `1000001` may not be accepted anymore, while previously it was.

If the team outputs `1000001.`, that shouldn't matter, because whether float-validation is used should only depend on the jury answer, not the team output.

Of course, the jury could update the model submission to always print `1000000.` instead, but they may never realize this issue exists in the first place.

For example, `print(1.0)` in python gives `1.0`, but `cout << 1.0;` in C++ gives just `1` (at least by default, on my machine/locale), so this feels like it's likely to cause bugs.

**Comment by Tagl** [comment]
> I would be careful to disable float comparison for integer-looking values: if the the floating-point output to some problems happens to be exactly `1000000`, then suddenly a team answer of `1000001` may not be accepted anymore, while previously it was.

I understand, but I think it never should have been accepted.

**Comment by RagnarGrootKoerkamp** [comment]
So with `float_tolerance: 0.01` and `answer: 1000000`, you would prefer to reject `1000001`?

I guess I can see your point: with mixed integer/float output, that does make sense.

But with float-only output, it feels like asking for issues.

Anyway no super strong opinion on this, but then I'd say that some sanity check like 'if `float_tolerance` is set, at least one period must appear in each answer file' is probably a good idea.

**Comment by Tagl** [comment]
Yes, I would prefer that because I can explicitly make the answer file have float tokens formatted with periods, which I always do already. Can also add another flag for these two behaviours.

**Comment by niemela** [comment]
Clarifying the expected behavior is good, but changing the expected behavior would be really bad. Of course, since the problem is that we have not specified it well enough, it's not completely clear what *is* currently expected. That said, to specify anything that is not what the current reference implementation does (i.e. problemtools) requires a very good reason *and* a strong belief that this has not been purposefully used (i.e. that the difference is very subtle.

If we want new behavior we should instead add new arguments, but let the existing ones keep their exact meaning.

**Comment by evouga** [comment]
I don't think there's any practical solution here other than the status quo, where the default validator treats integers as floating-point numbers when a floating-point tolerance is specified.

As Ragnar points out, C++ (and I assume many other languages) will print `double`s representing exact integers without a decimal point, and treating these as integers would break archived problem packages. It's also onerous if problem authors cannot use idiomatic prints in their solution code.

It might be useful for contest systems to warn the user when they try to install a problem containing integer judge data that will be treated as floating-point. But note that this situation is very common in current practice ("if a solution exists, print it. Otherwise print `-1`") and we don't need to dictate this behavior in the spec anyway.

**Comment by niemela** [comment]
> I don't think there's any practical solution here other than the status quo, where the default validator treats integers as floating-point numbers when a floating-point tolerance is specified.
> 
> As Ragnar points out, C++ (and I assume many other languages) will print `double`s representing exact integers without a decimal point, and treating these as integers would break archived problem packages. It's also onerous if problem authors cannot use idiomatic prints in their solution code.

Agreed!

> It might be useful for contest systems to warn the user when they try to install a problem containing integer judge data that will be treated as floating-point. But note that this situation is very common in current practice ("if a solution exists, print it. Otherwise print `-1`") and we don't need to dictate this behavior in the spec anyway.

I might be thinking wrong about this, but ... this only a problem if a token is "float-like" **and** a different token that is reasonably likely to happen by mistake is also "float-like" and within the tolerance **and** tolerance is set. I can't even imagine an example where those are all true. Can you?

I don't think it's a problem that **if** we have a problem with "if a solution exists, print it. Otherwise print `-1`" and `float_absolute_tolerance 0.2` that you could "trick" the system by printing `-1.1` instead of `-1`.

**Comment by evouga** [comment]
It could happen for a problem like, "given a set of points in the plane, print the number of points on the convex hull, and also the location of the centroid," where an off-by-one error in the first part would fit within the relative error tolerance.

In this case the problem needs a custom output validator that "knows" about the types of the two parts of the problem.

**Comment by Tagl** [comment]
For example a problem of finding line segment intersection where output is
which segments intersect and their intersection point. Then you have
integers and floats in same output. I think there are plenty of examples
where this comes up and even an issue on problemtools where this caught
someone by surprise.

**Comment by evouga** [comment]
Going back to `case_sensitive`: what should be the specification here?
- output is treated as encoded in UTF-8;
- for single-byte code points, the ASCII uppercase letters are considered equivalent to their corresponding lowercase letters;
- multi-byte code points must match exactly?

As an aside that's probably outside the scope of this issue: having `case_sensitive` that is (1) enabled by default and (2) only considers ASCII is begging for future pain.

**Comment by niemela** [comment]
Ok, @evouga, @Tagl, you're both right, I obviously have very bad imagination :).

In any case, I don't think we want to "fix" this, but I think it would make sense to add a warning in the documentation, saying something like "if you have both integer (or float I guess) values that you want to be exact and ones you want to have tolerance, you need to write a custom validator because the default validator will apply equally to all numerical values". The surprising part is of course not that if you want tolerance only for some float values you have to use a custom validator, but that integers are floats.

Does that make sense?

**Comment by niemela** [comment]
> for single-byte code points, unless `case_sensitive` is specified: the ASCII uppercase letters are considered equivalent to their corresponding lowercase letters;

I think this would break existing problems. We have problems with `åäö` that are expected to be equal to `ÅÄÖ` (from the Swedish alphabet) for example. I'm sure there are examples from the Icelandic alphabet as well, @Tagl?

**Comment by Tagl** [comment]
Yes, we have áðéíóúýþæö / ÁÐÉÍÓÚÝÞÆÖ

**Comment by niemela** [comment]
> Yes, we have áðéíóúýþæö / ÁÐÉÍÓÚÝÞÆÖ

Yes, but I think there might actually also be existing problems that use these, right?

**Comment by evouga** [comment]
...so what should we specify is the behavior of the comparison?

When it comes to `strcasecmp`, the POSIX spec unhelpfully says,

> In the POSIX locale, strcasecmp() and strncasecmp() shall behave as if the strings had been converted to lowercase and then a byte comparison performed. The results are unspecified in other locales.

**Comment by RagnarGrootKoerkamp** [comment]
Given that case insensitivity is usually best-effort, can't we leave it somewhat unspecified anyway? And then just refer to the docs of whatever the current default output validator calls? (I'm assuming that's the function you linked?)

**Comment by evouga** [comment]
We can. But this seems... non-ideal? As a contestant, I'd like it to be well-defined whether `ÁbC` will get AC when the problem specification says I must print `abc`. But maybe this is an intractable problem.

**Comment by niemela** [comment]
> When it comes to `strcasecmp`, the POSIX spec unhelpfully says,
> 
> > In the POSIX locale, strcasecmp() and strncasecmp() shall behave as if the strings had been converted to lowercase and then a byte comparison performed. The results are unspecified in other locales.

The POSIX (or "C") locale is arguably what we should be using. And I think it is what we **are**using

> We can. But this seems... non-ideal? As a contestant, I'd like it to be well-defined whether `ÁbC` will get AC when the problem specification says I must print `abc`. But maybe this is an intractable problem.

I think it *is* well defined? We should say that we are using the "POSIX"/"C" locale. That said, the docs I'm finding seems to say that this is what you (@evouga) suggested earlier, i.e. the "POSIX" locale will only lowercase ASCII?

So, one of the following is true:
- I'm wrong about what the "POSIX" locale does
- I'm wrong about the current implementation using "POSIX"
- I'm wrong about there actually being cases where we use `lowercase("ÅÄÖÄ) = åäö`
- I'm just **very** confused.
I don't think it's the last one...

**Comment by RagnarGrootKoerkamp** [comment]
> We have problems with åäö that are expected to be equal to ÅÄÖ (from the Swedish alphabet) for example

> Yes, we have áðéíóúýþæö / ÁÐÉÍÓÚÝÞÆÖ

Do these currently work? If so, is it because they are encoded as an ascii byte followed by a combining diacritic? So that upper/lower casing still works? I imagine that's more tricky for the non-ascii icelanic characters.

Maybe they work when the judge has the locale set to icelandic?

**Comment by Tagl** [comment]
> Do these currently work?

So far we have avoided non-ASCII output as we were originally advised to not go outside the standard ASCII for I/O.
We are trying out UTF-8 this year in contests, but none of our problems have had icelandic characters in output yet.
I will test it, but I assume the answer is no, it doesn't work.

I think it is reasonable that case-insensitivity only supports ASCII, but it would certainly be *nice* if the default validator had some utf-8 options as well. In general, I think UTF-8 in I/O is rare enough to warrant a custom output validator should these issues come up and we maybe shouldn't set too high expectations of what the default validator can handle. I was working on adding UTF-8 support to a very old project at my previous job so I'd be willing to try to add support for some of these without breaking backwards compatibility.

From the contestant side of things, I always just try to match the output directly, barring a trailing whitespace.

**Comment by thorehusfeldt** [comment]
1. Here is a Kattis problem using øæå: https://dpop22.kattis.com/contests/dpop22/problems/dpop22.rikke . `Sukkerstanger` in Sample 2 should certainly be rejected. 
2. Upper/lowercase is a can of worms and we might consider avoiding it in non-UTF8 problems. For instance, the correct upper case of the German word “Straße” (street) is “STRASSE”. I kid you not. Upper case of “Maße” (measurements) is (“MASSE”), and lowercase of “MASSE” is “masse”. So upper- and lowercase are not inverses. And this is just for German.

**Comment by evouga** [comment]
Closing as done.
